### ISSUE TITLE
Support webpack-cli options and/or webpack multiple configuration export

### ISSUE THREAD

**Issue by DBosley** [webpack/webpack-dev-middleware#312]
> The Issues page for this repository is not a support forum. **Please ask questions on StackOverflow or the webpack Gitter (https://gitter.im/webpack/webpack). Questions will be closed.** If you proceed with this form, please fill out **_all_** fields, or your issue may be closed as "invalid." Please do not delete this template from a new issue. Please remove this header to acknowledge this message.

Operating System: Ubuntu 18.04
Node Version: 8.11.1
NPM Version: 6.0.0
webpack Version: 4.12.0
webpack-hot-client Version: 4.0.3

- [x] This is a **feature** request
- [ ] This is a **bug**

### What is the motivation and/or use-case for the feature?
Webpack supports multiple configurations in a single file by exporting an array.
[Info Here](https://webpack.js.org/configuration/configuration-types/#exporting-multiple-configurations)
A new feature that was added recently (but currently lacks total documentation) is that you can set a 'name' property in your webpack config and pass `--config-name` to the cli to choose a specific config to build if you don't want to build all of them.

This feature request is to allow support of this functionality in the webpack-dev-middleware and related plugins.

Implementing this could be done in a couple ways:
- Support passing webpack-cli options into the middleware somehow
- Support choosing a config-name in the middleware options

I think the first solution is going to be the best approach if it's possible. This is because it would allow any webpack-cli arguments to be used with middlewares. If for some reason there's no way to go down this route, the second option would suffice for my use case, but it might be a good idea to look at other options listed [here](https://webpack.js.org/api/cli/) to see if they would merit being included as well

**Comment by shellscape** [comment]
Thanks for opening an issue :beer:

> Support passing webpack-cli options into the middleware somehow

`webpack-cli` is a mess of ad-hoc options added over time, so this would not be preferable. `webpack-command` is a vastly superior CLI, but even supporting all of the flags/options available on that would be bananas.

> Support choosing a config-name in the middleware options

This would actually be much more preferable, as it's a targeted, specific use option. The important distinction here between a module like this and something like a CLI is that this module is intended to be used as API directly. We should only ever add only the options we need.

Overall I think this is a really reasonable feature request. I haven't looked into how complex the change would be, but it should be doable.

**Comment by DBosley** [comment]
This issue arose while using `webpack-serve` and the way I envisioned the end-result of the implementation would be to pass the cli arguments in when running `webpack-serve` in the cli. Somehow allowing these to pass them through to whatever webpack-cli you're using behind the scenes. At the same time, I assume `webpack-serve` and `webpack-dev-middleware` might not be using any sort of webpack-cli under the hood at all, instead opting to use the webpack API, which is the reason I suggested the two approaches.

If either webpack-command or webpack-cli are actually being called in the middleware, passing whatever arguments through sounds fairly trivial. But if you're only using the API, it makes perfect sense to target specific use cases.

**Comment by shellscape** [comment]
Yeah take a quick peek at `package.json` and that'll tell you if a CLI is being used under the hood 😉 

The `webpack-serve` CLI use-case is interesting. In that situation, you'd first need support for a `configName` option in this repo, and then request that `webpack-serve` support a `--config-name` flag.

**Comment by ev1stensberg** [comment]
@shellscape This would align well in a separate repo that has a options file that both `serve` and `cli` can use. The feature wouldn't be hard to implement. Import the file and object assign on the file

**Comment by shellscape** [comment]
@ev1stensberg what? no. that would be a horrible architectural decision.

**Comment by ev1stensberg** [comment]
Perhaps, but it's better than directly importing the config and depending on that. A hard copy in Dev server would make it harder to maintain features if/when they are removed, added or modified

**Comment by shellscape** [comment]
I don't think you understand the conversation here. No one is talking about importing configs. 

@DBosley I'll see about getting into this next week. It may not look exactly like the suggestions you had, but pretty sure I know how we can make this happen.

**Comment by ev1stensberg** [comment]
I do, config -> yargs config. Makes it possible for us to bridge the libraries better. But seems like you got this under control.

**Comment by evilebottnawi** [comment]
@DBosley can you provide example how you can see this implementation?

**Comment by hinell** [comment]
Probably a bit late to party.
There are typically always two types of servers: development and productions ones.
In both you can specify which config to use by simply importing it and referencing it.

**Comment by eragon512** [comment]
hey, I'm a new contributor and I'd like to take up this issue

**Comment by evilebottnawi** [comment]
Yes, PR welcome

**Comment by EslamHiko** [comment]
@evilebottnawi I have a suggestion we can create 2 options one is `configs` or `configPath` its value will be the path to a config file ex:`webpack.middleware.js` or an array with a group of options with `mode` value. and we have the second option will be `mode` which if it's set it'll import the options from the group of options. so what do you think?

**Comment by evilebottnawi** [comment]
@EslamHiko this issue about supporting `name` as argument for build, webpack configuration can has multiple configuration (example https://github.com/webpack/webpack/tree/master/examples/multi-compiler), you developer provide this option we should compile only configuration with provided name
